automated-testing/ui-testing/readme.md

```diff
@@ -0,0 +1,64 @@
+
+Designing the UI tests around the functional tests makes sense.  For example, given an input form, a UI test would ensure that the visual representation is consistent across devices, is accessible and interactable, and is consistent across code changes.
+
+UI Tests will catch 'runtime' bugs that unit and functional tests won't.  For example if the submit button for an input form is rendered but not clickable due to a positioning bug in the UI, then this could be considered a runtime bug that would not have been caught by unit or functional tests.
+
+UI Tests can run on mock data or snapshots of production data, like in QA or staging.
+
+### Writing Tests
+
+Good UI tests follow a few general principles:
+
+- Choose a UI testing framework that enables quick feedback and is easy to use
+- Design the UI to be easily testable.  For example, add CSS selectors or set the id on elements in a web page to allow easier selecting.
+- Test on all primary devices that the user uses, don't just test on a single device or OS.
+- When a test mutates data ensure that data is created on demand and cleaned up after.  The consequence of not doing this would be inconsistent testing.  
+
+### Gotchas
+
+UI Testing can get very challenging at the lower level, especially with a testing framework like Selenium.  If you choose to go this route, then you'll likely encounter timeouts, missing elements, and you'll have significant friction with the testing framework itself.  Due to many issues with UI testing there have been a number of free and paid solutions that help alleviate certain issues with frameworks like Selenium.  This is why you'll find Cypress in the recommended frameworks as it solves many of the known issues with Selenium.
+
+This is an important point though.  Depending on the UI testing framework you choose will result in either a smoother test creation experience, or a very frustrating and time consuming one.  If you were to choose just Selenium the development costs and time costs would likely be very high.  It's better to use either a framework built on top of Selenium or one that attempts to solve many of the problems with something like Selenium.
+
+Note there that there are further considerations as when running headlessly the UI can render differently than what you may see on your development machine, particularly with web applications.  Furthermore, note that when rendering in different page dimensions elements may disappear on the page due to CSS rules, therefore not be selectable by certain frameworks with default options out of the box.  All of these issues can be resolved and worked around, but the rendering demonstrates another particular challenge of UI testing.
+
+## Specific Guidance
+
+Recommended testing frameworks:
+
+- Web:
+  - BrowserStack: <https://www.browserstack.com>
+  - Cypress: <https://www.cypress.io>
+  - Jest: <https://jestjs.io/docs/en/snapshot-testing>
+  - Selenium: <https://www.selenium.dev>
```

Comment:
Prefer Puppeteer or Playwright here before Selenium

Reply:
Also: https://testing-library.com/ is recommended for framework testing.

Reply:
Added!